sources/api/corndog/src/main.rs

```diff
@@ -104,23 +109,83 @@ where
     path
 }
 
-/// Applies the requested sysctls to the system.  The keys are used to generate the appropriate
-/// path, and the value its contents.
-fn set_sysctls<K>(sysctls: HashMap<K, String>)
+/// Validates sysctl key/value pairs for systemd-sysctl compatibility
+fn validate_sysctl_entry<K>(key: K, value: &str) -> Result<()>
+where
+    K: AsRef<str>,
+{
+    let key = key.as_ref();
+
+    // Basic validation of key format
+    if !key.chars().all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '_' || c == '-' || c == '*') {
+        return error::InvalidSysctlKeySnafu { key: key.to_string() }.fail();
+    }
+
+    // Ensure value doesn't contain newlines which could break the config file format
+    if value.contains('\n') {
+        return error::InvalidSysctlValueSnafu {
+            key: key.to_string(),
+            value: value.to_string()
+        }.fail();
+    }
+
+    Ok(())
+}
+
+/// Restarts systemd-sysctl service
+fn restart_sysctl() -> Result<()> {
+    let status = Command::new("systemctl")
+        .args(["restart", "systemd-sysctl"])
+        .status()
+        .context(error::SystemctlSnafu)?;
+
+    if !status.success() {
+        return error::SystemctlFailedSnafu.fail();
+    }
+
+    Ok(())
+}
+
+/// Applies the requested sysctls to the system using systemd-sysctl.
+fn set_sysctls<K>(sysctls: HashMap<K, String>) -> Result<()>
 where
     K: AsRef<str>,
 {
+    // TODO: make the priority configurable?
+    const SYSCTL_CONFIG: &str = "/etc/sysctl.d/95-corndog.conf";
+
+    // Validate all entries first
+    for (key, value) in &sysctls {
+        validate_sysctl_entry(key, value)?;
+    }
+
+    // Create config string
+    let mut config = String::new();
     for (key, value) in sysctls {
-        let key = key.as_ref();
-        let path = sysctl_path(key);
-        if let Err(e) = fs::write(path, value) {
-            // We don't fail because sysctl keys can vary between kernel versions and depend on
-            // loaded modules.  It wouldn't be possible to deploy settings to a mixed-kernel fleet
-            // if newer sysctl values failed on your older kernels, for example, and we believe
-            // it's too cumbersome to have to specify in settings which keys are allowed to fail.
-            error!("Failed to write sysctl value '{}': {}", key, e);
-        }
+        config.push_str(&format!("{}={}\n", key.as_ref(), value));
     }
+
+    // Append to file, creating it if it doesn't exist
+    let mut file = fs::OpenOptions::new()
+        .create(true)
+        .append(true)
+        .open(SYSCTL_CONFIG)
+        .context(error::OpenSysctlConfigSnafu)?;
+
+    file.write_all(config.as_bytes())
+        .context(error::WriteSysctlConfigSnafu)?;
+
+    // Restart systemd-sysctl to apply changes
+    let status = Command::new("systemctl")
+        .args(["restart", "systemd-sysctl"])
+        .status()
+        .context(error::SystemctlSnafu)?;
```

Comment:
You don't really need to restart the whole sysctld service.
```
/usr/lib/systemd/systemd-sysctl /etc/sysctl.d/99-corndog.conf
```
Would apply this specific file to sysctld.

And could you wrap the binary path and args into constants? Something like below and put it along with the definition of the `SYSCTL_CONFIG_PATH`.
```
const SYSTEMD_SYSCTL_BIN: &str = "/usr/lib/systemd/systemd-sysctl";
```

Reply:
I'll need to confirm this, but I believe only applying the specific file may not necessarily respect the ordering of the rest of the config files, so `sysctl restart systemd-sysctl` is correct IMO.

For example, the Cilium sysctl.conf file has a prefix of `99-zzz`, so applying the `95` priority corndog file by itself may clobber any entries that are also set in the Cilium conf file.

If we kept `sysctl restart systemd-ctl`, would we still want these wrapped into constants?